File: app/controllers/Application.scala
package controllers

import models.PahoClientModel
import utils.ScalaMD5Sum
import play.api._
import play.api.mvc._
import play.api.libs.iteratee.Enumerator
import play.api.libs.iteratee.Iteratee
import play.api.libs.iteratee.Concurrent
import play.api.Logger._
import play.api.libs.concurrent.Execution.Implicits._
import scala.util.Properties

object Application extends Controller {
  
  def index = Action {
    Ok(views.html.index("Your new application is ready."))
  }

  def mqtt() = WebSocket.using[String] { request =>
  //println("In websocket action.")
    val md5pass =ScalaMD5Sum.computeSum(Properties.envOrElse("THINGFABRIC_PASSWORD", "" ))//"a029d0df84eb5549c641e04a9ef389e5"

    val (out,channel) = Concurrent.broadcast[String]

    val mqtt = new PahoClientModel(Properties.envOrElse("THINGFABRIC_M2M_ENDPOINT", "q.m2m.io" ), 1883, java.util.UUID.randomUUID.toString.substring(0,22), Properties.envOrElse("THINGFABRIC_USERNAME", ""), md5pass, true, 30, channel)

    val in = Iteratee.foreach[String](x => {
      //logger.info("Websocket msg: " + x)
      try {
        val Array(cmd, topic, qos, message) = x.split("<--")
        cmd match {
          case "SUBSCRIBE" => mqtt.subscribe(Properties.envOrElse("THINGFABRIC_M2M_DATA_CHANNEL", "public/thingfabric/#" ), qos.toInt)
          case "UNSUBSCRIBE" => mqtt.unsubscribe(topic)
          case "PUBLISH" => mqtt.publish(topic, message, qos.toInt)
          case "CHECK" => if(mqtt.client.isConnected()) channel.push("-->CONNECTED")
          case "PING" => mqtt.publish(topic, message, 0)
          case _ => logger.info("Unknown cmd: " + x)
        }
      } catch {
        case e: Exception => {
          logger.error("Exception occured in websocket", e)
        }
      }
    }).mapDone { x =>
      mqtt.disconnect
    }

    (in, out)
  }
  
}

File: app/controllers/Application.java
package controllers;

import play.*;
import play.mvc.*;

import views.html.*;
import models.PahoClientModel;
import utils.ScalaMD5Sum;
import akka.actor.ActorRef;
import akka.actor.Cancellable;
import akka.actor.Props;
import play.libs.Akka;
import play.libs.F.Callback0;
import play.libs.F.Callback;
import org.eclipse.paho.client.mqttv3.*;
import java.util.UUID;
import java.security.NoSuchAlgorithmException;

public class Application extends Controller {

    public static Result index() {
        return ok(index.render("Your new application is ready."));
    }

    public static WebSocket<String> mqtt() {
        return new WebSocket<String>() {
            public void onReady(WebSocket.In<String> in, final WebSocket.Out<String> out) {
                try{
                    String md5pass = "";
                    String username = "";
                    if (System.getenv().containsKey("THINGFABRIC_USERNAME") && System.getenv().containsKey("THINGFABRIC_PASSWORD")){
                        md5pass = ScalaMD5Sum.computeSum(System.getenv().get("THINGFABRIC_PASSWORD")) ;
                        username =  System.getenv().get("THINGFABRIC_USERNAME");
                    }
                    final PahoClientModel mqtt = new PahoClientModel(System.getenv("THINGFABRIC_M2M_ENDPOINT"), 1883, UUID.randomUUID().toString().substring(0,22), username, md5pass, true, 30, out);

                    in.onMessage(new Callback<String>() {
                        public void invoke(String event) throws MqttException{
                            Logger.info(event);
                            String[] arr = event.split("<--");
                            String cmd = arr[0];
                            String topic = arr[1];
                            Logger.info(cmd);
                            if (cmd.equals("SUBSCRIBE")) mqtt.subscribe(System.getenv().get("THINGFABRIC_M2M_DATA_CHANNEL"), 1);
                            if (cmd.equals("UNSUBSCRIBE")) mqtt.unsubscribe(topic);
                            if (cmd.equals("PUBLISH") ) mqtt.publish(System.getenv().get("THINGFABRIC_M2M_DATA_CHANNEL"), arr[3], 1);
                            if (cmd.equals("CHECK"))  if(mqtt.client.isConnected()) out.write("-->CONNECTED");
                            if (cmd.equals("PING")) mqtt.publish(topic, arr[3], 0);

                        }
                    });

                    in.onClose(new Callback0() {
                        public void invoke() throws Throwable {
                            mqtt.disconnect();
                        }
                    });
                } catch (MqttSecurityException se){
                     Logger.info("MqttSecurityException");
                } catch (MqttException e){
                    Logger.info("MqttException");
                } catch (NoSuchAlgorithmException ex){
                    Logger.info("NoSuchAlgorithmException");
                }
            }

        };
    }
}
